Realm/RLMRealm.mm

```diff
@@ -120,23 +124,98 @@ static bool shouldForciblyDisableEncryption() {
 @implementation RLMRealm {
     NSHashTable<RLMFastEnumerator *> *_collectionEnumerators;
     bool _sendingNotifications;
+
+    // Needs to be atomic because the timeout callback is called on the main
+    // thread and not the RLMRealm's thread, so we could try to end the task
+    // from multiple threads at once
+    std::atomic<NSUInteger> _backgroundTaskIdentifier;
+    bool _addedBackgroundNotification;
 }
 
 + (BOOL)isCoreDebug {
     return realm::Version::has_feature(realm::feature_Debug);
 }
 
+id g_sharedApplication;
 + (void)initialize {
     static bool initialized;
     if (initialized) {
         return;
     }
     initialized = true;
 
+#if REALM_IOS
+    // Using NSClassFromString rather than directly referencing UIApplication
+    // avoids the need to link UIKit.framework, and we can't actually write
+    // `[UIApplication sharedApplication]` due to compiling as extension-safe
+    g_sharedApplication = [NSClassFromString(@"UIApplication") sharedApplication];
+
+    // Background tasks aren't supported if we're running inside an extension
+    if (![g_sharedApplication respondsToSelector:@selector(beginBackgroundTaskWithExpirationHandler:)]) {
+        g_sharedApplication = nil;
+    }
+#endif
+
     RLMCheckForUpdates();
     RLMSendAnalytics();
 }
 
+#if REALM_IOS
+// On iOS we want to wrap our write transactions in background tasks so that the
+// OS doesn't suspend the app while we hold the write lock, as that will block
+// any other apps in the same app group from using the Realm. Doing this
+// unconditionally is extremely slow, so instead we only do it if we're already
+// in the background, and just register for the background state transition
+// notification otherwise
+- (void)beginBackgroundTask {
+    if (!g_sharedApplication) {
+        return;
+    }
+
+    if ([g_sharedApplication applicationState] == UIApplicationStateBackground) {
+        [self applicationDidEnterBackground:nil];
+        return;
+    }
+
+    [NSNotificationCenter.defaultCenter addObserver:self
+                                           selector:@selector(applicationDidEnterBackground:)
+                                               name:@"UIApplicationDidEnterBackground"
+                                             object:nil];
+    _addedBackgroundNotification = true;
+}
+
+- (void)applicationDidEnterBackground:(NSNotification *)notification {
+    if (notification && !self.inWriteTransaction) {
```

Comment:
we're using this approach and actually it's easy to receive `Realm accessed from incorrect thread` exception when trying to cancel a transaction in the write state.